File: sdks/php/tests/Unit/ClaimsTest.php
<?php

declare(strict_types=1);

namespace Hearth\Tests\Unit;

use DateTimeImmutable;
use Hearth\Claims;
use PHPUnit\Framework\TestCase;

/**
 * Unit tests for Claims — the typed JWT claims accessor.
 *
 * Tests are written before the implementation per the TDD contract.
 * Run with: vendor/bin/phpunit tests/Unit/ClaimsTest.php
 */
final class ClaimsTest extends TestCase
{
    private function makeClaims(array $overrides = []): Claims
    {
        return new Claims(array_merge([
            'sub'         => 'usr_abc123',
            'iss'         => 'https://auth.example.com',
            'aud'         => ['my-client'],
            'exp'         => time() + 3600,
            'iat'         => time() - 60,
            'jti'         => 'jwt-id-xyz',
            'scope'       => 'openid profile email',
            'token_type'  => 'access',
            'roles'       => ['admin', 'editor'],
            'permissions' => ['users:read', 'users:write'],
            'groups'      => ['engineering', 'backend'],
            'oid'         => 'org_tenant42',
            'org_groups'  => ['/tenant42/engineering'],
            'sid'         => 'sess_abc',
            'tid'         => 'realm_xyz',
        ], $overrides));
    }

    public function testSubjectReturnsSubClaim(): void
    {
        self::assertSame('usr_abc123', $this->makeClaims()->subject());
    }

    public function testIssuerReturnsIssClaim(): void
    {
        self::assertSame('https://auth.example.com', $this->makeClaims()->issuer());
    }

    public function testAudiencesNormalisesStringToArray(): void
    {
        $claims = $this->makeClaims(['aud' => 'single-aud']);
        self::assertSame(['single-aud'], $claims->audiences());
    }

    public function testAudiencesReturnsArrayUnchanged(): void
    {
        self::assertSame(['my-client'], $this->makeClaims()->audiences());
    }

    public function testExpiryReturnsDateTimeImmutable(): void
    {
        $exp    = time() + 3600;
        $claims = $this->makeClaims(['exp' => $exp]);
        $expiry = $claims->expiry();
        self::assertInstanceOf(DateTimeImmutable::class, $expiry);
        self::assertSame($exp, $expiry->getTimestamp());
    }

    public function testExpiryReturnsNullWhenAbsent(): void
    {
        $claims = $this->makeClaims(['exp' => null]);
        self::assertNull($this->makeClaims(array_diff_key([], ['exp' => '']))->expiry());
        // Inline null test
        self::assertNull((new Claims([]))->expiry());
    }

    public function testIssuedAtReturnsDateTimeImmutable(): void
    {
        $iat    = time() - 120;
        $claims = $this->makeClaims(['iat' => $iat]);
        $issuedAt = $claims->issuedAt();
        self::assertInstanceOf(DateTimeImmutable::class, $issuedAt);
        self::assertSame($iat, $issuedAt->getTimestamp());
    }

    public function testJwtIdReturnsJtiClaim(): void
    {
        self::assertSame('jwt-id-xyz', $this->makeClaims()->jwtID());
    }

    public function testJwtIdReturnsEmptyStringWhenAbsent(): void
    {
        self::assertSame('', (new Claims([]))->jwtID());
    }

    public function testScopeReturnsSpaceDelimitedString(): void
    {
        self::assertSame('openid profile email', $this->makeClaims()->scope());
    }

    public function testScopesReturnsParsedArray(): void
    {
        self::assertSame(['openid', 'profile', 'email'], $this->makeClaims()->scopes());
    }

    public function testHasScopeReturnsTrueForPresent(): void
    {
        self::assertTrue($this->makeClaims()->hasScope('profile'));
    }

    public function testHasScopeReturnsFalseForAbsent(): void
    {
        self::assertFalse($this->makeClaims()->hasScope('admin:write'));
    }

    public function testHasRoleReturnsTrueForPresent(): void
    {
        self::assertTrue($this->makeClaims()->hasRole('admin'));
    }

    public function testHasRoleReturnsFalseForAbsent(): void
    {
        self::assertFalse($this->makeClaims()->hasRole('superuser'));
    }

    public function testHasRoleReturnsFalseWhenClaimAbsent(): void
    {
        self::assertFalse((new Claims([]))->hasRole('admin'));
    }

    public function testHasPermissionReturnsTrueForPresent(): void
    {
        self::assertTrue($this->makeClaims()->hasPermission('users:read'));
    }

    public function testHasPermissionReturnsFalseForAbsent(): void
    {
        self::assertFalse($this->makeClaims()->hasPermission('billing:write'));
    }

    public function testInGroupReturnsTrueForPresent(): void
    {
        self::assertTrue($this->makeClaims()->inGroup('engineering'));
    }

    public function testInGroupReturnsFalseForAbsent(): void
    {
        self::assertFalse($this->makeClaims()->inGroup('sales'));
    }

    public function testInOrgReturnsTrueForExactMatch(): void
    {
        self::assertTrue($this->makeClaims()->inOrg('org_tenant42'));
    }

    public function testInOrgReturnsFalseForMismatch(): void
    {
        self::assertFalse($this->makeClaims()->inOrg('org_other'));
    }

    public function testInOrgReturnsFalseWhenClaimAbsent(): void
    {
        self::assertFalse((new Claims([]))->inOrg('org_any'));
    }

    public function testTokenTypeReturnsTokenTypeClaim(): void
    {
        self::assertSame('access', $this->makeClaims()->tokenType());
    }

    public function testOrganizationIdReturnsOidClaim(): void
    {
        self::assertSame('org_tenant42', $this->makeClaims()->organizationId());
    }

    public function testOrganizationIdReturnsNullWhenAbsent(): void
    {
        self::assertNull((new Claims([]))->organizationId());
    }

    public function testOrgGroupsReturnsOrgGroupsClaim(): void
    {
        self::assertSame(['/tenant42/engineering'], $this->makeClaims()->orgGroups());
    }

    public function testOrgGroupsReturnsEmptyArrayWhenAbsent(): void
    {
        self::assertSame([], (new Claims([]))->orgGroups());
    }

    public function testGetReturnsRawClaimValue(): void
    {
        self::assertSame('sess_abc', $this->makeClaims()->get('sid'));
    }

    public function testGetReturnsNullForAbsentClaim(): void
    {
        self::assertNull($this->makeClaims()->get('nonexistent'));
    }
}

File: sdks/kotlin/hearth-core/src/test/kotlin/io/hearth/sdk/ClaimsTest.kt
package io.hearth.sdk

import com.nimbusds.jwt.JWTClaimsSet
import java.time.Instant
import java.util.Date
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ClaimsTest {

    private fun makeClaims(block: JWTClaimsSet.Builder.() -> JWTClaimsSet.Builder): Claims {
        val builder = JWTClaimsSet.Builder()
            .subject("user-123")
            .issuer("https://auth.example.com")
            .audience("my-app")
            .expirationTime(Date.from(Instant.now().plusSeconds(3600)))
            .issueTime(Date.from(Instant.now()))
        val claimsSet = builder.block().build()
        return Claims(claimsSet)
    }

    @Test
    fun `subject returns sub claim`() {
        val c = makeClaims { this }
        assertEquals("user-123", c.subject())
    }

    @Test
    fun `issuer returns iss claim`() {
        val c = makeClaims { this }
        assertEquals("https://auth.example.com", c.issuer())
    }

    @Test
    fun `audiences returns aud list`() {
        val c = makeClaims { this }
        assertEquals(listOf("my-app"), c.audiences())
    }

    @Test
    fun `scope and scopes parse correctly`() {
        val c = makeClaims { claim("scope", "read write admin") }
        assertEquals("read write admin", c.scope())
        assertEquals(listOf("read", "write", "admin"), c.scopes())
        assertTrue(c.hasScope("write"))
        assertFalse(c.hasScope("delete"))
    }

    @Test
    fun `hasRole returns true when role present`() {
        val c = makeClaims { claim("roles", listOf("admin", "viewer")) }
        assertTrue(c.hasRole("admin"))
        assertFalse(c.hasRole("superuser"))
    }

    @Test
    fun `hasRole returns false when roles claim absent`() {
        val c = makeClaims { this }
        assertFalse(c.hasRole("admin"))
    }

    @Test
    fun `hasPermission returns true when permission present`() {
        val c = makeClaims { claim("permissions", listOf("user:read", "user:write")) }
        assertTrue(c.hasPermission("user:read"))
        assertFalse(c.hasPermission("user:delete"))
    }

    @Test
    fun `hasPermission returns false when permissions claim absent`() {
        val c = makeClaims { this }
        assertFalse(c.hasPermission("user:read"))
    }

    @Test
    fun `inGroup returns true when group present`() {
        val c = makeClaims { claim("groups", listOf("engineering", "platform")) }
        assertTrue(c.inGroup("engineering"))
        assertFalse(c.inGroup("marketing"))
    }

    @Test
    fun `get returns raw claim value`() {
        val c = makeClaims { claim("custom_field", "custom_value") }
        assertEquals("custom_value", c.get("custom_field"))
    }

    @Test
    fun `get returns null for absent claim`() {
        val c = makeClaims { this }
        assertEquals(null, c.get("nonexistent"))
    }

    @Test
    fun `expiry and issuedAt return instants`() {
        val c = makeClaims { this }
        assertTrue(c.expiry().isAfter(Instant.now()))
        assertTrue(c.issuedAt().isBefore(Instant.now().plusSeconds(1)))
    }

    // ── §4 gaps: inOrg, tokenType, organizationId, orgGroups ──────────────────

    @Test
    fun `inOrg returns true when oid matches`() {
        val c = makeClaims { claim("oid", "org-abc") }
        assertTrue(c.inOrg("org-abc"))
    }

    @Test
    fun `inOrg returns false when oid does not match`() {
        val c = makeClaims { claim("oid", "org-abc") }
        assertFalse(c.inOrg("org-xyz"))
    }

    @Test
    fun `inOrg returns false when oid claim absent`() {
        val c = makeClaims { this }
        assertFalse(c.inOrg("org-abc"))
    }

    @Test
    fun `organizationId returns oid claim value`() {
        val c = makeClaims { claim("oid", "org-abc") }
        assertEquals("org-abc", c.organizationId())
    }

    @Test
    fun `organizationId returns null when oid absent`() {
        val c = makeClaims { this }
        assertEquals(null, c.organizationId())
    }

    @Test
    fun `tokenType returns token_type claim`() {
        val c = makeClaims { claim("token_type", "access") }
        assertEquals("access", c.tokenType())
    }

    @Test
    fun `tokenType returns empty string when claim absent`() {
        val c = makeClaims { this }
        assertEquals("", c.tokenType())
    }

    @Test
    fun `tokenType returns required_action when token is a required-action token`() {
        val c = makeClaims { claim("token_type", "required_action") }
        assertEquals("required_action", c.tokenType())
    }

    @Test
    fun `orgGroups returns org_groups claim list`() {
        val c = makeClaims { claim("org_groups", listOf("/org-slug/eng", "/org-slug/platform")) }
        assertEquals(listOf("/org-slug/eng", "/org-slug/platform"), c.orgGroups())
    }

    @Test
    fun `orgGroups returns empty list when claim absent`() {
        val c = makeClaims { this }
        assertEquals(emptyList<String>(), c.orgGroups())
    }
}
